Support units in binary operations

danielhuppmann opened the pull request:
# Please confirm that this PR has done the following:

- [x] Tests Added
- [x] Documentation Added
- ~Name of contributors Added to AUTHORS.rst~
- [x] Description in RELEASE_NOTES.md Added

# Description of PR

This PR changes the implementation of the binary ops to support units (via iam-units and pint).

Default behaviour:
 - Cast values and units to a `pint.Quantity` using the **iam-units** registry and perform operations with pint unit handling (`ignore_units=False`)
    **Warning**: By casting units to pints and then performing operations, the return-format may look different.
   In some of the tests, `EJ/yr` is transformed to `exajoule / year` (in full format) or `EJ / a` (in compact format, implemented here). This is over-ridden for some methods (addition, multiplication, division) if all units are identical.
 - Passing a pint quantity as argument will work, e.g, `df.add("Primary Energy", Quantity(2, "EJ/yr"), "new")`.
 - **pint** will raise appropriate errors if it doesn't know how to handle the units.
 - In line with common practice in the IAM community, a dimensionless quantity is marked as unit `""` (empty string) rather than the word `dimensionless` (this works in a round-trip to-from-to pint).

Alternative over-ride:
 - A user can set `ignore_units=True`, in which case the operations will work just on the values and set `unit` to "unknown" (per suggestion by @znicholls in https://github.com/IAMconsortium/pyam/issues/536#issuecomment-847569955) or to the value of `ignore_units` (if it's a string).

Known issues:
- pandas fillna does not work with non-numeric values (in particular not with pint.Quantity).

Side note:
 - I simplified the test for the `apply()` function.

To-dos in later PRs (it's already quite a beast):
 - Make the fillna feature work with pint.
 - Add `ignore_units` to the `apply()` function.
 - Support custom pint registry instances.
 - Pass unit formatting options to pint as kwargs.

closes #537 
closes #535

pjuergens commented on pyam/_ops.py:
```diff
@@ -70,14 +111,33 @@ def _op_data(df, name, method, axis, fillna=None, args=(), **kwds):
+    if ignore_units is False:
+        rename_args = ("dimensionless", "")
+        _value = pd.DataFrame(
+            [[i.magnitude, str(i.units).replace(*rename_args)] for i in result.values],
```
Pint can return the unit as `EJ / yr` instead of `exajoule / year` with this code snipplet
```python
# format the the unit as string 'EJ / yr' 
unit = '{:~}'.format(c.units)
```
To implement it, the line should read something like
```python
[[i.magnitude, '{:~}'.format(str(i.units)).replace(*rename_args)] for i in result.values],
```
I'm not sure, if the `str()` is still neded then and how it returns dimensionless units

danielhuppmann replied:
Implemented.

pjuergens commented:
So it's just a comment and not a full review. I hope some of the others have time for an in-depth-review, cause I'd really like to use this feature :)

danielhuppmann commented:
Thanks for the suggestion, @pjuergens - implemented it and updated the unit tests, because shorter units are generally preferable, I think. Also updated the description of this PR to reflect this change.

However, note that the underlying issue remains, because a user will generally not receive the same notation for a unit - in our case, `EJ / yr` is now returned as `EJ / a`. I don't see that as a problem, just something to be aware of...

znicholls commented on pyam/_ops.py:
```diff
@@ -1,24 +1,39 @@
+    therefore a pint.Quantity is transformed to a pd.Series of quantities.
+    """
+    if isinstance(a, Quantity):
+        return pd.Series([a] * len(b), index=b.index), b
```
We should probably add explicit tests for this.

[Pint-pandas](https://github.com/hgrecco/pint-pandas) is an attempt to support something like this. At the moment I don't think it works for pyam because the entire series has to have the same unit, but maybe something to keep in mind for future.

znicholls reviewed (commented):
Looks good to me. My major concern would be the casting of each individual value to/from pint.Quantity. That looks like it could be an expensive operation as soon as the data is a moderate size. We could test that, or we could just not worry about it for now.

danielhuppmann commented:
Thanks for the comment @znicholls! I had a look at [pint-pandas](https://github.com/hgrecco/pint-pandas), but as you write, this assumes that an entire column in a dataframe has the same unit - this is not something that pyam satisfies, not even on a per-variable level. A user could currently have a scenario ensemble where "Primary Energy" is in EJ for some scenarios and "GWh" in others. Something to discuss for a pyam v2.0 release...

About speed and casting, I'd follow the strategy I was taught by @gidden - first make it work, then make it fast...

znicholls commented:
> About speed and casting, I'd follow the strategy I was taught by @gidden - first make it work, then make it fast...

It's a great strategy!

pjuergens commented on pyam/_ops.py:
```diff
@@ -46,13 +61,39 @@ def _op_data(df, name, method, axis, fillna=None, args=(), **kwds):
+            df, axis, value, cols, key
+        )
+
+    # fast-pass on units: override pint for some methods if all kwds have same the unit
```
```suggestion
    # fast-pass on units: override pint for some methods if all kwds have the same unit
```

pjuergens commented on pyam/core.py:
```diff
@@ -1838,13 +1849,16 @@ def subtract(self, a, b, name, axis="variable", fillna=None, append=False):
+    ):
         """Compute the product of timeseries data between `a` and `b` along an `axis`
 
         This function computes `a * b`. If `a` or `b` are lists, the method applies
```
Should we add a note in the description here and for the divide-function that units might be reformatted by pint? For addition and subtraction this should not be necessary.
Also I was wondering which unit pint returns when multiplying e.g. `PJ` with `1 / PJ`. - Does it recognise it as `dimensionless` or will it return `PJ / PJ`? This might become important when using more complicated functions in `apply` - so we can keep it in mind to test in a following PR.

danielhuppmann replied:
Added a docstring to all four binary-ops functions and `apply()` - after all, as soon as not all units are identical, all functions will revert to abbreviated-units style of pint.

About more elaborate operations corner cases like you describe, my experience is that pint is quite good at dealing with those...

pjuergens replied:
Addition and subtraction should lead to an error, if it's not done with identical units.

I think you are right about pint that it recognizes `PJ / PJ`. What it doesn't recognize are such things like `TWh / PJ` - but I would leave that to the user to convert it into `dimensionsless`.

znicholls replied:
> Addition and subtraction should lead to an error, if it's not done with identical units.

Why? If you add PJ and EJ or Mt CO2 and Gt CO2, the units are compatible so why should it error? 

> What it doesn't recognize are such things like `TWh / PJ`

Pint is generally fine with this. E.g. if I had W * s / J then you could convert to dimensionless and it should all work without an issue. Whether pint knows about TWh or not is a different question (is TWh defined in pint or iam_units?)

pjuergens replied:
> Why? If you add PJ and EJ or Mt CO2 and Gt CO2, the units are compatible so why should it error?

I haven't thought about that :D I was thinking about different units like EJ and Mt CO2 and pint should raise and error then, hopefully ;-)

> Pint is generally fine with this. E.g. if I had W * s / J then you could convert to dimensionless and it should all work without an issue. Whether pint knows about TWh or not is a different question (is TWh defined in pint or iam_units?)

I just tried pint with meters and inch and it returned meter/inch, which I had to convert (using pint) into dimensionless later. But I don't see an issue with that - users shout convert the units either before or after by themselves.

pjuergens reviewed (changes requested):
Basically just two minor comments concerning documentation. As far as I can tell the coding looks good to me. Thanks for the work!

pjuergens reviewed (approved):
In my opinion it's good to go :)
